lib/src/fortio.cpp

```diff
@@ -0,0 +1,337 @@
+#include <cstdint>
+#include <cstdio>
+#include <cstring>
+#include <limits>
+#include <stdexcept>
+
+#include <ecl/fortio.h>
+
+#ifdef HAVE_NETINET_IN_H
+    #include <netinet/in.h>
+#elif HAVE_ARPA_INET_H
+    #include <arpa/inet.h>
+#elif HAVE_WINSOCK2_H
+    #include <winsock2.h>
+#endif
+
+namespace {
+
+/*
+ * Functions in fortio will try and roll back as much state as possible in case
+ * of errors, to provide some exception safety. Since these error paths are
+ * more cleanly handled with returned, the rollback is tied to the fguard's
+ * destructor.
+ *
+ * However, if the rollback itself fails, the file is left in an unspecific
+ * state, but we have no immediate way to report that, as the destructor does
+ * not return a value. Instead, an exception is thrown, and all function use
+ * function-try-blocks. While it's generally bad practice to throw from
+ * destructors, the eclfio functions are extern C'd, and no exceptions leak.
+ * Since there are no other destructors run from these functions than that of
+ * fguard, the destructor-exception is ok.
+ */
+struct fguard {
+    FILE* fp;
+    std::fpos_t pos;
+
+    fguard( FILE* x ) : fp( x ) {
+        const auto err = std::fgetpos( fp, &this->pos );
+
+        if( err ) {
+            /*
+             * when fgetpos fails, presumably everything is broken and we
+             * didn't even record our previous position. no point even trying
+             * to roll back from this, so set fp to nullptr and bail
+             */
+            this->fp = nullptr;
+            throw std::runtime_error("");
+        }
+    }
+
+    ~fguard() noexcept( false ) {
+        if( !this->fp ) return;
+
+        const auto err = std::fsetpos( fp, &this->pos );
+        if( err ) throw std::runtime_error("");
+    }
```

Comment:
I know this is a controversial topic, but I wouldn't `throw` from a destructor 
Specifically in this case, if you are unable to seek, it means something went awfully wrong before you tried to seek. And doing it like this will actually mask the real error.

Take for instance `eclfio_sizeof`. What can fail there is the `std::fread`, and if that fails, the method is supposed to return `ECL_ERR_READ`. But if `std::fread` failed so badly that the following seek fails as well, then the method returns `ECL_ERR_SEEK`. So there is a discrepancy between the error reported and what error actually occurred

Reply:
I disagree - if you're in a situation where the *rollback* fails, literally nothing can be guaranteed anymore, and it doesn't matter if it was the read that failed, or an argument was invalid, because at that point callers must take *extra* care to restore a well-known state.

Maybe the return type of ECL_ERR_SEEK is a red herring, because it's not simply a seek that's failed probably, it's the exception guarantee to boot. It's pretty disastrous (and rare, I would imagine). But I would still argue that I don't if the read fail if that also means that my seeking is busted.